web/app/js/components/Tap.jsx

```diff
@@ -0,0 +1,142 @@
+import _ from 'lodash';
+import ErrorBanner from './ErrorBanner.jsx';
+import PageHeader from './PageHeader.jsx';
+import PropTypes from 'prop-types';
+import React from 'react';
+import { withContext } from './util/AppContext.jsx';
+import { Button, Form, Icon, Input } from 'antd';
+import './../../css/tap.css';
+
+class Tap extends React.Component {
+  static propTypes = {
+    pathPrefix: PropTypes.string.isRequired
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      errors: "",
+      resource: "",
+      namespace: "",
+      messages: [],
+      maxLinesToDisplay: 40,
+      webSocketRequestSent: false
+    };
+  }
+
+  componentWillUnmount() {
+    this.stopTapStreaming();
+  }
+
+  onWebsocketRecv = e => {
+    this.setState({
+      messages: _(this.state.messages)
+        .push(e.data)
+        .takeRight(this.state.maxLinesToDisplay)
+        .value()
+    });
+  }
+
+  onWebsocketClose = e => {
+    if (!e.wasClean) {
+      this.setState({
+        errors: `Websocket [${e.code}] ${e.reason}`
+      });
+    }
+    this.stopTapStreaming();
+  }
+
+  onWebsocketOpen = () => {
+    this.ws.send(JSON.stringify({
+      id: "tap-web",
+      resource: this.state.resource,
+      namespace: this.state.namespace
+    }));
+    this.setState({
+      webSocketRequestSent: true
+    });
+  }
+
+  startTapSteaming() {
+    this.setState({
+      messages: []
+    });
+
+    let protocol = window.location.protocol === "https:" ? "wss" : "ws";
+    let tapWebSocket = `${protocol}://${window.location.host}${this.props.pathPrefix}/api/tap`;
+
+    this.ws = new WebSocket(tapWebSocket);
+    this.ws.onmessage = this.onWebsocketRecv;
+    this.ws.onclose = this.onWebsocketClose;
+    this.ws.onopen = this.onWebsocketOpen;
+  }
+
+  stopTapStreaming() {
+    this.setState({
+      webSocketRequestSent: false
+    });
+
+    this.ws.close();
```

Comment:
Am pretty sure this method shouldn't call `ws.close()`, since it's called from `onWebsocketClose`, which is called after `ws.close()` is called, creating a loop. Basically, when somebody clicks the Stop button, I think that triggers the following method calls:

```
handleTapStop => stopTapStreaming => ws.close() => onWebsocketClose => stopTapStreaming
```

Whereas I think it would be better to rewrite as:

```
handleTapStop => ws.close => onWebsocketClose => stopTapStreaming
              => stopTapStreaming
```

That would require removing this `ws.close()` call, and instead calling `ws.close()` directly from `handleTapStop`:

```
  handleTapStop = () => {
    this.ws.close(1000);
    this.stopTapStreaming();
  }
```

And have the `stopTapStreaming` method only handle updating state and clearing the interval.

Reply:
I think you should also set a 1000 status code in the `handleTapStop` method, to indicate that this was an expected closure. It looks like the default is 1005. See also https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close#Parameters and my comments in api_handlers.go.

Reply:
ahhhh yeah thanks!